WordPress/gutenberg
# Allow longhand and shorthand properties in `theme.json` and block attributes

Related: https://github.com/WordPress/gutenberg/issues/31337, https://github.com/WordPress/gutenberg/pull/31585
~~Depends on: https://github.com/WordPress/gutenberg/pull/31483~~

## Description

This PR:
- Flattens the [configuration array used to map CSS property names](https://github.com/WordPress/gutenberg/pull/31641/files#diff-b03597cc3da199e5aa5a94950e8241135904853e7c3b82d758e42744878afae7R198) with paths to the appropriate value in theme.json and block attributes
- Updates border and spacing support so that border radius, margin and padding support allow both shorthand and longhand styles e.g. `margin: 10px` or `margin-left: 20px`
- Updates the theme.json and block support style tests to cover shorthand and longhand styles
- Updates global styles renderer to handle shorthand/longhand style values and configuration

The driving force behind this was achieving the refined border support UI in https://github.com/WordPress/gutenberg/issues/31337. To do so, the border block support needs to allow custom values for each corner's border radius. This PR addresses this need while also maintaining backwards compatibility for a single border radius value.

## How has this been tested?
Automated tests.

Theme.json stylesheet generation: 
`/var/www/html/wp-content/plugins/gutenberg/phpunit/class-wp-theme-json-test.php`

Block support inline styles: 
`packages/block-editor/src/hooks/test/style.js`
`packages/edit-site/src/components/editor/test/global-styles-renderer.js`

Manual testing:
1. Apply this PR
2. Enable custom margin and padding support via theme.json
3. Add theme styles for the `core/group` block in the theme.json as well. Try simple shorthand margin and padding values
4. Open the site editor and in the global styles sidebar switch to the `By block type` tab, then expand the Group block panel.
5. Confirm the margin and padding controls work and their global style values can be updated and saved correctly
6. Edit the theme.json margin / padding styles to use split values per side
7. Reload the site editor and confirm the group block's margin/padding controls still function
8. Create a post within the normal block editor and add a group block
9. Select the group block and configure its margin and paddding
10. Save and confirm the spacing displays correctly.

_This PR does not update the border support UI, to test the border changes manually, try https://github.com/WordPress/gutenberg/pull/31585_

1. Checkout out the #31585 branch `update/border-block-support-ui`
    (that's all that is needed, it contains the changes from this PR)
2. Enable border support UI via theme.json - [example theme.json gist](https://gist.github.com/aaronrobertshaw/4deded861032467d4dccfe0797d3befd)
3. Build (you might need to run `npm install`)
4. Load editor, create post, add group block, and select it
5. Confirm border support panel and radius control display and function as expected

## Screenshots 
Via use of these changes in https://github.com/WordPress/gutenberg/pull/31585

<img src="https://user-images.githubusercontent.com/60436221/117418310-9cf4af00-af5e-11eb-9cd5-3f7e5655b2f9.gif" width="400" />

## Types of changes
New feature.

## Checklist:
